src/lua/alloc.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2010-2024, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "lua/utils.h"
+
+#include <lauxlib.h> /* luaL_error */
+
+#define LUA_MEMORY_MIN (256ll * 1024 * 1024)
+#define LUA_MEMORY_DEFAULT (2ll * 1024 * 1024 * 1024)
+
+/**
+ * The default allocator function which is wrapped into a new one.
+ */
+static lua_Alloc orig_alloc;
+/**
+ * The maximum allowed LuaJIT memory limit.
+ */
+static size_t memory = LUA_MEMORY_DEFAULT;
+/**
+ * Currently allocated memory amount by LuaJIT.
+ */
+static size_t allocated;
+
+/**
+ * Lua custom memory allocation function. It extends the original one with
+ * a memory counter and a limit check.
+ */
+static void *
+luaT_alloc(void *ud, void *ptr, size_t osize, size_t nsize)
+{
+	size_t new_allocated = allocated + nsize - osize;
+	if (new_allocated > memory) {
+		diag_set(OutOfMemory, nsize, "luaT_alloc", "Lua objects");
+		return NULL;
+	}
+	allocated = new_allocated;
+
+	return orig_alloc(ud, ptr, osize, nsize);
```

Comment:
I suppose that `orig_alloc` may return `NULL` too. So, we should check the returned pointer and update allocated only if it is not `NULL`.

Reply:
You're absolutely right, thank you for finding this point.

Reply:
I'm not sure, that we should set the `diag` here. That may allow us to separate the error from our allocator and internal ones. @Totktonada, what do you think?

Reply:
@georgiy-belyanin Any updates here?

Reply:
I no longer do `diag_set` there. By default lua handles this as `not enough memory`. I suppose it's quite ok since this way the error is quite clear, no extra error handling logic is provided.

Reply:
Seems OK for me, thanks for the update!